docs/proposals/locking.md

```diff
@@ -108,3 +108,198 @@ support:
 |11 |Reject a push containing a binary file currently locked by someone else|pre-receive hook on server, allow --force to override (i.e. existing parameter to git push)
+
+## Implementation details
+### Types
+To make the implementing locking on the lfs-test-server as well as other servers
+in the future easier, it makes sense to create a `lock` package that can be
+depended upon from any server. This will go along with Steve's refactor which
+touches the `lfs` package quite a bit.
+
+Below are enumerated some of the types that will presumably land in this
+sub-package.
+
+```go
+// Lock represents a single lock that against a particular path.
+//
+// Locks returned from the API may or may not be currently active, according to
+// the Expired flag.
+type Lock struct {
+        // Path is an absolute path to the file that is locked as a part of this
+        // lock.
+        Path string `json:"path"`
+        // Creator is the author who initiated this lock.
+        Creator struct {
+               Name  string `json:"name"`
+               Email string `json:"email"`
+        } `json:"creator"`
+        // CreatedAt is a required parameter that represents the instant in time
+        // that this lock was created. For most server implementations, this
+        // should be set to the instant at which the lock was initially
+        // received.
+        CreatedAt time.Time `json:"created_at"`
+        // ExpiresAt is an optional parameter that represents the instant in
+        // time that the lock stopped being active. If the lock is still active,
+        // the server can either a) not send this field, or b) send the
+        // zero-value of time.Time.
+        ExpiresAt time.Time `json:"expired_at,omitempty"`
+}
+
+// Active returns whether or not the given lock is still active against the file
+// that it is protecting.
+func (l *Lock) Active() bool {
+        return time.IsZero(l.ExpiresAt)
+}
+```
+
+### Proposed Commands
+
+#### `git lfs lock <path>`
+
+The `lock` command will be used in accordance with the multi-branch flow as
+proposed above to request that lock be granted to the specific path passed an
+argument to the command.
+
+```go
+// LockRequest encapsulates the payload sent across the API when a client would
+// like to obtain a lock against a particular path on a given remote.
+type LockRequest struct {
+        // Path is the path that the client would like to obtain a lock against.
+        Path      string `json:"path"`
+        // Remote is the remote on which the client would like to obtain the
+        // lock.
+        Remote    string `json:"remote"`
+        // LatestRemoteCommit is the SHA of the last known commit from the
+        // remote that we are trying to create the lock against, as found in
+        // `.git/refs/origin/<name>`.
+        LatestRemoteCommit string `json:"latest_remote_commit"`
```

Comment:
I think @technoweenie brings up a good point in [this](https://github.com/github/git-lfs/pull/1223#discussion_r63512395).